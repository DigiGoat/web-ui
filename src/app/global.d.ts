import * as Bootstrap from 'bootstrap';

declare global {
  interface Window {
    gtag: ((command: 'config' | 'event' | 'set' | 'js', parameter: string | Date, arguments?: Record<string, string | boolean> | string) => void) | never;
    clarity: ((command: 'set', parameter: string, arguments?: Record<string, string> | string) => void) | never;
  }
}

declare module 'bootstrap' {
  const bootstrap: typeof Bootstrap;
}
