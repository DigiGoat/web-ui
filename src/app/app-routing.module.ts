import { NgModule, type OnInit } from '@angular/core';
import { TitleStrategy as NgTitleStrategy, Route, RouterModule } from '@angular/router';

import type { Observable } from 'rxjs';
import { kiddingScheduleGuard } from './guards/kidding-schedule/kidding-schedule.guard';
import { referencesGuard } from './guards/references/references.guard';
import { BucksComponent } from './pages/bucks/bucks.component';
import { DoesComponent } from './pages/does/does.component';
import { HomeComponent } from './pages/home/home.component';
import { KiddingScheduleComponent } from './pages/kidding-schedule/kidding-schedule.component';
import { NotFoundComponent } from './pages/not-found/not-found.component';
import { ReferencesComponent } from './pages/references/references.component';
import { TitleStrategy } from './strategies/title.strategy';

const routes: Route[] = [
  { path: '', component: HomeComponent, title: 'Home' },
  {
    path: 'does', children: [
      { path: '', component: DoesComponent, title: 'Does' },
      { path: ':goat', component: DoesComponent, title: ':goat - Does' }
    ]
  }, {
    path: 'bucks', children: [
      { path: '', component: BucksComponent, title: 'Bucks' },
      { path: ':goat', component: BucksComponent, title: ':goat - Bucks' }
    ]
  },
  {
    path: 'references', canMatch: [referencesGuard], children: [
      { path: '', component: ReferencesComponent, title: 'References' },
      { path: ':goat', component: ReferencesComponent, title: ':goat - References' }
    ]
  },
  {
    path: 'kidding-schedule', canMatch: [kiddingScheduleGuard], children: [
      { path: '', component: KiddingScheduleComponent, title: 'Kidding Schedule' },
      { path: ':goat', component: KiddingScheduleComponent, title: ':goat - Kidding Schedule' }
    ]
  },
  {
    path: '**', component: NotFoundComponent, title: 'Not Found'
  }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule],
  providers: [{ provide: NgTitleStrategy, useClass: TitleStrategy }]
})
export class AppRoutingModule { }

export interface Page extends OnInit {
  setDescription(): Observable<void> | void;
}
