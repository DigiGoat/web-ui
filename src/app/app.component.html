<!-- Navbar to display tabs at the top of the page -->
<header class="navbar navbar-expand-lg sticky-top color-scheme-secondary">
  <div class="container-fluid">
    <img ngSrc="./assets/icons/android-chrome-192x192.png" alt="Logo" class="d-inline-block align-text-top" width="70" height="70" priority>
    <a class="navbar-brand ps-2 text-center text-lg-start" routerLink="/">
      <p class="h4" [innerHTML]="splitString(app.menubarTitle)"></p>
    </a>
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent"
      aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0 nav nav-underline nav-fill w-100">
        <li class="nav-item">
          <a class="nav-link" routerLinkActive="active" aria-current="page" routerLink="/" [routerLinkActiveOptions]="{exact: true}"
            data-bs-toggle="collapse" data-bs-target=".navbar-collapse.show">Home</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" routerLinkActive="active" aria-current="page" routerLink="/does" data-bs-toggle="collapse"
            data-bs-target=".navbar-collapse.show">Does</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" routerLinkActive="active" aria-current="page" routerLink="/bucks" data-bs-toggle="collapse"
            data-bs-target=".navbar-collapse.show">Bucks</a>
        </li>
      </ul>
    </div>
  </div>
</header>
<br>
<!-- Put in minimum height div to put footer at bottom of page -->
<div class="px-3 py-3" style="min-height: 100vh; max-width: 100vw !important;">
  <router-outlet></router-outlet>
</div>
<br>
<!-- Footer with link to source code -->
<footer class="bd-footer color-scheme-secondary">
  <br>
  <div class="text-center"> {{app.owner}} <br>
    <a [href]="'mailto:' + app.email">{{app.email}}</a>
  </div><br>
  <div class="container  text-center" style="min-height: 5vh;">This Site Was Created and Is Maintained Using <a
      href="https://github.com/DigiGoat" target="_blank">DigiGoat</a>
    <br>See The Source Code On <a href="https://github.com/DigiGoat/web-ui" target="_blank">Github</a>
  </div>
  <br>
</footer>
