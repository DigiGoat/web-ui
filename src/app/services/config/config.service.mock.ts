import type { ConfigService } from './config.service';

export const ConfigServiceMock = {
  homeTitle: 'TEST_TITLE',
  owner: 'TEST_OWNER',
  email: 'TEST_EMAIL',
  homeDescription: 'TEST_DESCRIPTION',
  menubarTitle: 'TEST_MENUBAR_TITLE',
  tabTitle: 'TEST_TAB_TITLE',
  link: 'TEST_LINK'
} as ConfigService;
export const EmptyConfigServiceMock = {
  homeTitle: '',
  owner: '',
  email: '',
  homeDescription: '',
  menubarTitle: '',
  tabTitle: '',
  link: ''
} as ConfigService;
