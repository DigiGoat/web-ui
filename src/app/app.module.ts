import { NgOptimizedImage } from '@angular/common';
import { provideHttpClient, withFetch } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { BrowserModule, provideClientHydration } from '@angular/platform-browser';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';
import { ImgDirective } from './directives/img/img.directive';
import { GoatCardComponent } from './elements/goat-card/goat-card.component';
import { GoatsComponent } from './elements/goats/goats.component';
import { GoatCardComponent as GoatCardModalComponent } from './elements/modal/goat-card/goat-card.component';
import { ModalComponent } from './elements/modal/modal/modal.component';
import { NotFoundComponent as NotFoundModalComponent } from './elements/modal/not-found/not-found.component';
import { BucksComponent } from './pages/bucks/bucks.component';
import { DoesComponent } from './pages/does/does.component';
import { HomeComponent } from './pages/home/home.component';
import { NotFoundComponent } from './pages/not-found/not-found.component';
import { AgePipe } from './pipes/age.pipe';
import { CarouselComponent } from './elements/carousel/carousel.component';


@NgModule({
  declarations: [
    AppComponent,
    BucksComponent,
    DoesComponent,
    GoatCardComponent,
    GoatsComponent,
    HomeComponent,
    AgePipe,
    NotFoundComponent,
    ModalComponent,
    GoatCardModalComponent,
    NotFoundModalComponent,
    ImgDirective,
    CarouselComponent
  ],
  imports: [
    BrowserModule,
    NgOptimizedImage,
    AppRoutingModule
  ],
  providers: [
    provideClientHydration(),
    provideHttpClient(withFetch())
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
