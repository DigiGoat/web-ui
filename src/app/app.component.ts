import { Component } from '@angular/core';

import { ConfigService } from './services/config/config.service';


@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss']
})
export class AppComponent {
  constructor(public app: ConfigService) { }
  splitString(string: string) {
    const splitString = string.split(' ');
    const half = Math.floor(splitString.length / 2);
    const firstHalf = splitString.slice(0, half).join(' ');
    const secondHalf = splitString.slice(half).join(' ');
    return `${firstHalf}<br>${secondHalf}`;
  }
}
