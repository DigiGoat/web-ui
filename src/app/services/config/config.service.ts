import { Injectable } from '@angular/core';

import config from '../../../assets/resources/config.json';


@Injectable({
  providedIn: 'root'
})
export class ConfigService {
  private config: Record<string, string | boolean | Record<string, string | Record<string, string>>> = config;
  get title(): string {
    if (this.config['title']) {
      return this.config['title'] as string;
    } else if (this.config['homeTitle']) {
      return this.config['homeTitle'] as string;
    } else if (this.config['menubarTitle']) {
      return this.config['menubarTitle'] as string;
    }
    return '';
  }
  get shortTitle(): string {
    if (this.config['shortTitle']) {
      return this.config['shortTitle'] as string;
    } else if (this.config['tabTitle']) {
      return this.config['tabTitle'] as string;
    }
    return '';
  }
  get owner(): string {
    if (this.config['owner']) {
      return this.config['owner'] as string;
    }
    return '';
  }
  get email(): string {
    if (this.config['email']) {
      return this.config['email'] as string;
    }
    return '';
  }
  get homeDescription(): string {
    if (this.config['homeDescription']) {
      return this.config['homeDescription'] as string;
    }
    return '';
  }
  get link(): string {
    if (this.config['link']) {
      return `${this.config['link']}${(this.config['link'] as string).endsWith('/') ? '' : '/'}`;
    }
    return '';
  }
  get analytics(): Analytics {
    if (this.config['analytics']) {
      return this.config['analytics'] as Analytics;
    }
    return {};
  }
  get colors(): ColorScheme {
    if (this.config['colors']) {
      return this.config['colors'] as ColorScheme;
    }
    return {};
  }
  get kiddingSchedule(): boolean {
    if (this.config['kiddingSchedule']) {
      return this.config['kiddingSchedule'] as boolean;
    }
    return false;
  }
  get kiddingScheduleDescription(): string {
    if (this.config['kiddingScheduleDescription']) {
      return this.config['kiddingScheduleDescription'] as string;
    }
    return '';
  }
  get socials(): Socials {
    if (this.config['socials']) {
      return this.config['socials'] as Socials;
    }
    return {};
  }
  get references(): boolean {
    if (this.config['references']) {
      return this.config['references'] as boolean;
    }
    return false;
  }
  get forSale(): boolean {
    if (this.config['forSale']) {
      return this.config['forSale'] as boolean;
    }
    return false;
  }
  get saleTerms(): string {
    if (this.config['saleTerms']) {
      return this.config['saleTerms'] as string;
    }
    return '';
  }
  get firebase(): Firebase | false {
    if (this.config['firebase'] && typeof this.config['firebase'] === 'object') {
      // Make sure alll firebase keys are present
      if (!this.config['firebase']['apiKey'] || !this.config['firebase']['projectId'] || !this.config['firebase']['messagingSenderId'] || !this.config['firebase']['appId']) {
        console.error('Missing Firebase keys');
        return false;
      }
      const config = this.config['firebase'] as Firebase;
      config.authDomain = config.authDomain || `${config.projectId}.firebaseapp.com`;
      config.storageBucket = config.storageBucket || `${config.projectId}.firebasestorage.app`;
      return config;
    }
    return false;
  }
}

type Analytics = { gtag?: string, clarity?: string; };
type ColorScheme = {
  background?: 'wood';
  main?: string;
  secondary?: string;
  tertiary?: string;
  quaternary?: string;
  light?: {
    main?: string;
    secondary?: string;
    tertiary?: string;
    quaternary?: string;
  };
};
type Socials = { facebook?: string, instagram?: string, threads?: string; };
type Firebase = { apiKey?: string, authDomain?: string, projectId?: string, storageBucket?: string, messagingSenderId?: string, appId?: string; };
