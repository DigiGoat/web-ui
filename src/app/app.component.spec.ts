import { ComponentFixture, TestBed } from '@angular/core/testing';

import { AppComponent } from './app.component';
import { ConfigService } from './services/config/config.service';
import { ConfigServiceMock, EmptyConfigServiceMock } from './services/config/config.service.mock';

jest.mock('./services/config/config.service');
describe('AppComponent', () => {
  let component: AppComponent;
  let fixture: ComponentFixture<AppComponent>;
  let html: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [AppComponent],
      providers: [ConfigService]
    })
      .compileComponents();

    fixture = TestBed.createComponent(AppComponent);
    component = fixture.componentInstance;
    html = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  it('should split strings in half', () => {
    expect(component.splitString('Test String')).toBe('Test<br>String');
    expect(component.splitString('Really Long Test String')).toBe('Really Long<br>Test String');
    expect(component.splitString('Odd Test String')).toBe('Odd<br>Test String');
    expect(component.splitString('Test   String')).toBe('Test <br> String');
    expect(component.splitString('Test')).toBe('Test');
  });
  describe('With a Config (v1.0.0)', () => {
    beforeEach(() => {
      component.config = ConfigServiceMock;
      fixture.detectChanges();
    });
    it('should have a menubar title', () => {
      const element = html.querySelector('[test-id=menubar]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('TEST_MENUBAR_TITLE');
    });
    it('should have an owner', () => {
      const element = html.querySelector('[test-id=owner]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('TEST_OWNER');
    });
    it('should have an email', () => {
      const element = html.querySelector<HTMLAnchorElement>('[test-id=email]');
      expect(element).toBeTruthy();
      expect(element?.innerHTML).toBe('TEST_EMAIL');
      expect(element?.href).toBe('mailto:TEST_EMAIL');
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
  describe('Without a Config', () => {
    beforeEach(() => {
      component.config = EmptyConfigServiceMock;
      fixture.detectChanges();
    });
    it('should not have a menubar title', () => {
      const element = html.querySelector('[test-id=menubar]');
      expect(element?.innerHTML).toBeFalsy();
    });
    it('should have an owner', () => {
      const element = html.querySelector('[test-id=owner]');
      expect(element?.innerHTML).toBeFalsy();
    });
    it('should have an email', () => {
      const element = html.querySelector<HTMLAnchorElement>('[test-id=email]');
      expect(element?.innerHTML).toBeFalsy();
      expect(element?.href).toBe('mailto:');
    });
    it('should match the snapshot', () => {
      expect(fixture).toMatchSnapshot();
    });
  });
});
