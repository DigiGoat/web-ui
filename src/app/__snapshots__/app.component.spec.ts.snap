// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent With a Config (v1.0.0) should match the snapshot 1`] = `
<app-root
  config={[Function Object]}
>
  <header
    class="navbar navbar-expand-lg sticky-top color-scheme-secondary"
  >
    <div
      class="container-fluid"
    >
      <img
        alt="Logo"
        class="d-inline-block align-text-top"
        height="70"
        ngsrc="./assets/icons/android-chrome-192x192.png"
        priority=""
        width="70"
      />
      <a
        class="navbar-brand ps-2 text-center text-lg-start"
        routerlink="/"
      >
        <p
          class="h4"
          test-id="menubar"
        >
          TEST_MENUBAR_TITLE
        </p>
      </a>
      <button
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-bs-target="#navbarSupportedContent"
        data-bs-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul
          class="navbar-nav me-auto mb-2 mb-lg-0 nav nav-underline nav-fill w-100"
        >
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="nav-link"
              data-bs-target=".navbar-collapse.show"
              data-bs-toggle="collapse"
              routerlink="/"
              routerlinkactive="active"
            >
              Home
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="nav-link"
              data-bs-target=".navbar-collapse.show"
              data-bs-toggle="collapse"
              routerlink="/does"
              routerlinkactive="active"
            >
              Does
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="nav-link"
              data-bs-target=".navbar-collapse.show"
              data-bs-toggle="collapse"
              routerlink="/bucks"
              routerlinkactive="active"
            >
              Bucks
            </a>
          </li>
        </ul>
      </div>
    </div>
  </header><br /><div
    class="px-3 py-3"
    style="min-height: 100vh; max-width: 100vw !important;"
  >
    <router-outlet />
  </div><br /><footer
    class="bd-footer color-scheme-secondary"
  >
    <br />
    <div
      class="text-center"
    >
      <span
        test-id="owner"
      >
        TEST_OWNER
      </span>
      <br />
      <a
        href="mailto:TEST_EMAIL"
        test-id="email"
      >
        TEST_EMAIL
      </a>
    </div>
    <br />
    <div
      class="container text-center"
      style="min-height: 5vh;"
    >
      This Site Was Created and Is Maintained Using 
      <a
        href="https://github.com/DigiGoat"
        target="_blank"
      >
        DigiGoat
      </a>
      <br />
      See The Source Code On 
      <a
        href="https://github.com/DigiGoat/web-ui"
        target="_blank"
      >
        Github
      </a>
    </div>
    <br />
  </footer>
</app-root>
`;

exports[`AppComponent Without a Config should match the snapshot 1`] = `
<app-root
  config={[Function Object]}
>
  <header
    class="navbar navbar-expand-lg sticky-top color-scheme-secondary"
  >
    <div
      class="container-fluid"
    >
      <img
        alt="Logo"
        class="d-inline-block align-text-top"
        height="70"
        ngsrc="./assets/icons/android-chrome-192x192.png"
        priority=""
        width="70"
      />
      <a
        class="navbar-brand ps-2 text-center text-lg-start"
        routerlink="/"
      >
        <p
          class="h4"
          test-id="menubar"
        />
      </a>
      <button
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
        class="navbar-toggler"
        data-bs-target="#navbarSupportedContent"
        data-bs-toggle="collapse"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
        id="navbarSupportedContent"
      >
        <ul
          class="navbar-nav me-auto mb-2 mb-lg-0 nav nav-underline nav-fill w-100"
        >
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="nav-link"
              data-bs-target=".navbar-collapse.show"
              data-bs-toggle="collapse"
              routerlink="/"
              routerlinkactive="active"
            >
              Home
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="nav-link"
              data-bs-target=".navbar-collapse.show"
              data-bs-toggle="collapse"
              routerlink="/does"
              routerlinkactive="active"
            >
              Does
            </a>
          </li>
          <li
            class="nav-item"
          >
            <a
              aria-current="page"
              class="nav-link"
              data-bs-target=".navbar-collapse.show"
              data-bs-toggle="collapse"
              routerlink="/bucks"
              routerlinkactive="active"
            >
              Bucks
            </a>
          </li>
        </ul>
      </div>
    </div>
  </header><br /><div
    class="px-3 py-3"
    style="min-height: 100vh; max-width: 100vw !important;"
  >
    <router-outlet />
  </div><br /><footer
    class="bd-footer color-scheme-secondary"
  >
    <br />
    <div
      class="text-center"
    >
      <span
        test-id="owner"
      >
      </span>
      <br />
      <a
        href="mailto:"
        test-id="email"
      >
      </a>
    </div>
    <br />
    <div
      class="container text-center"
      style="min-height: 5vh;"
    >
      This Site Was Created and Is Maintained Using 
      <a
        href="https://github.com/DigiGoat"
        target="_blank"
      >
        DigiGoat
      </a>
      <br />
      See The Source Code On 
      <a
        href="https://github.com/DigiGoat/web-ui"
        target="_blank"
      >
        Github
      </a>
    </div>
    <br />
  </footer>
</app-root>
`;
